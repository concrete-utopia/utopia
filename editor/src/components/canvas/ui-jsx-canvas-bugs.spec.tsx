import {
  testCanvasRender,
  testCanvasRenderInline,
  testCanvasRenderInlineMultifile,
} from './ui-jsx-canvas.test-utils'

describe('UiJsxCanvas', () => {
  it('#747 - DOM object constructor cannot be called as a function', () => {
    testCanvasRender(
      null,
      `
import React from 'react'
import Utopia, {
  Scene,
  Storyboard,
  registerModule,
} from 'utopia-api'

const DefaultComments = [
  {
    userName: 'forbidden_one',
    contents: 'Integer eu imperdiet enim. Aenean vitae sem et ex feugiat accumsan et a mi.',
  },
]

const Comment = () => <div data-uid='comment-root'>hat</div>

export var App = () =>
  true ? DefaultComments.map((comment) => <Comment comment={comment} />) : null

export var storyboard = (
  <Storyboard data-uid='sb'>
    <Scene
      data-uid='scene'
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
    >
      <App data-uid='app' />
    </Scene>
  </Storyboard>
)
    `,
    )
  })

  it('Supports in-scope variables with the same names as intrinsic components', () => {
    const result = testCanvasRenderInline(
      null,
      `
import React from 'react'
import Utopia, {
  Scene,
  Storyboard,
  registerModule,
} from 'utopia-api'

export const App = () => {
  const div = React.useRef()
  return <div data-uid='app-root' ref={div} />
}

export var storyboard = (
  <Storyboard data-uid='sb'>
    <Scene
      data-uid='scene'
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
    >
      <App data-uid='app' />
    </Scene>
  </Storyboard>
)
    `,
    )

    expect(result).toMatchInlineSnapshot(`
      "<div style=\\"all: initial\\">
        <div
          id=\\"canvas-container\\"
          style=\\"position: absolute\\"
          data-utopia-valid-paths=\\"sb sb/scene sb/scene/app sb/scene/app:app-root\\"
          data-utopia-root-element-path=\\"sb\\"
        >
          <div
            data-utopia-scene-id=\\"sb/scene\\"
            data-path=\\"sb/scene\\"
            style=\\"
              overflow: hidden;
              position: absolute;
              background-color: rgba(255, 255, 255, 1);
              box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
              background-image: linear-gradient(
                  to bottom left,
                  #e7e7e7 25%,
                  transparent 25%
                ),
                linear-gradient(to bottom left, transparent 75%, #e7e7e7 75%),
                linear-gradient(to bottom right, #e7e7e7 25%, transparent 25%),
                linear-gradient(to bottom right, transparent 75%, #e7e7e7 75%);
              background-size: 12px 12px, 12px 12px, 12px 12px, 12px 12px;
              background-position: -9px 0px, -3px -6px, 3px 6px, -3px 0;
              left: 0;
              top: 0;
              width: 375px;
              height: 812px;
            \\"
            data-uid=\\"scene\\"
          >
            <div data-uid=\\"app-root\\" data-path=\\"sb/scene/app:app-root\\"></div>
          </div>
        </div>
      </div>
      "
    `)
  })

  it('Handles importing default exports', () => {
    const result = testCanvasRenderInlineMultifile(
      null,
      `
import React from 'react'
import Utopia, {
  Scene,
  Storyboard,
  registerModule,
} from 'utopia-api'
import Appy from './app'

export var storyboard = (
  <Storyboard data-uid='sb'>
    <Scene
      data-uid='scene'
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
    >
      <Appy data-uid='app' />
    </Scene>
  </Storyboard>
)
`,
      {
        'app.js': `
import React from 'react'
export default function App(props) {
  return <div data-uid='app-outer-div'>
    <div data-uid='inner-div'>hello</div>
  </div>
}`,
      },
    )

    expect(result).toMatchInlineSnapshot(`
      "<div style=\\"all: initial\\">
        <div
          id=\\"canvas-container\\"
          style=\\"position: absolute\\"
          data-utopia-valid-paths=\\"sb sb/scene sb/scene/app\\"
          data-utopia-root-element-path=\\"sb\\"
        >
          <div
            data-utopia-scene-id=\\"sb/scene\\"
            data-path=\\"sb/scene\\"
            style=\\"
              overflow: hidden;
              position: absolute;
              background-color: rgba(255, 255, 255, 1);
              box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
              background-image: linear-gradient(
                  to bottom left,
                  #e7e7e7 25%,
                  transparent 25%
                ),
                linear-gradient(to bottom left, transparent 75%, #e7e7e7 75%),
                linear-gradient(to bottom right, #e7e7e7 25%, transparent 25%),
                linear-gradient(to bottom right, transparent 75%, #e7e7e7 75%);
              background-size: 12px 12px, 12px 12px, 12px 12px, 12px 12px;
              background-position: -9px 0px, -3px -6px, 3px 6px, -3px 0;
              left: 0;
              top: 0;
              width: 375px;
              height: 812px;
            \\"
            data-uid=\\"scene\\"
          >
            <div data-uid=\\"app-outer-div\\" data-path=\\"sb/scene/app\\">
              <div data-uid=\\"inner-div\\">hello</div>
            </div>
          </div>
        </div>
      </div>
      "
    `)
  })

  it('#1717 - Works with user components called Scene', () => {
    const result = testCanvasRenderInlineMultifile(
      null,
      `
import React from 'react'
import { Scene as SC, Storyboard } from 'utopia-api'
import App from './app'

export var Scene = (props) => {
  return (
    <div
      data-uid='same-file-app-div'
      data-label='Scene Thing'
      style={{
        position: 'relative',
        width: '100%',
        height: '100%',
        backgroundColor: 'blue',
      }}
    />
  )
}

export var storyboard = (
  <Storyboard data-uid='storyboard-entity'>
    <SC
      data-label='Imported App'
      data-uid='scene-1-entity'
      style={{
        position: 'absolute',
        left: 0,
        top: 0,
        width: 375,
        height: 812,
      }}
    >
      <App data-uid='app-entity' />
    </SC>
    <SC
      data-label='Same File App'
      data-uid='scene-2-entity'
      style={{
        position: 'absolute',
        left: 400,
        top: 0,
        width: 375,
        height: 812,
      }}
    >
      <Scene data-uid='same-file-app-entity' />
    </SC>
  </Storyboard>
)
`,
      {
        'app.js': `
import React from 'react'
export default function App(props) {
  return <div data-uid='app-outer-div'>
    <div data-uid='inner-div'>hello</div>
  </div>
}`,
      },
    )

    expect(result).toMatchInlineSnapshot(`
      "<div style=\\"all: initial\\">
        <div
          id=\\"canvas-container\\"
          style=\\"position: absolute\\"
          data-utopia-valid-paths=\\"storyboard-entity storyboard-entity/scene-1-entity storyboard-entity/scene-1-entity/app-entity storyboard-entity/scene-2-entity storyboard-entity/scene-2-entity/same-file-app-entity storyboard-entity/scene-2-entity/same-file-app-entity:same-file-app-div\\"
          data-utopia-root-element-path=\\"storyboard-entity\\"
        >
          <div
            data-utopia-scene-id=\\"storyboard-entity/scene-1-entity\\"
            data-path=\\"storyboard-entity/scene-1-entity\\"
            style=\\"
              overflow: hidden;
              position: absolute;
              background-color: rgba(255, 255, 255, 1);
              box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
              background-image: linear-gradient(
                  to bottom left,
                  #e7e7e7 25%,
                  transparent 25%
                ),
                linear-gradient(to bottom left, transparent 75%, #e7e7e7 75%),
                linear-gradient(to bottom right, #e7e7e7 25%, transparent 25%),
                linear-gradient(to bottom right, transparent 75%, #e7e7e7 75%);
              background-size: 12px 12px, 12px 12px, 12px 12px, 12px 12px;
              background-position: -9px 0px, -3px -6px, 3px 6px, -3px 0;
              left: 0;
              top: 0;
              width: 375px;
              height: 812px;
            \\"
            data-uid=\\"scene-1-entity\\"
            data-label=\\"Imported App\\"
          >
            <div
              data-uid=\\"app-outer-div\\"
              data-path=\\"storyboard-entity/scene-1-entity/app-entity\\"
            >
              <div data-uid=\\"inner-div\\">hello</div>
            </div>
          </div>
          <div
            data-utopia-scene-id=\\"storyboard-entity/scene-2-entity\\"
            data-path=\\"storyboard-entity/scene-2-entity\\"
            style=\\"
              overflow: hidden;
              position: absolute;
              background-color: rgba(255, 255, 255, 1);
              box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
              background-image: linear-gradient(
                  to bottom left,
                  #e7e7e7 25%,
                  transparent 25%
                ),
                linear-gradient(to bottom left, transparent 75%, #e7e7e7 75%),
                linear-gradient(to bottom right, #e7e7e7 25%, transparent 25%),
                linear-gradient(to bottom right, transparent 75%, #e7e7e7 75%);
              background-size: 12px 12px, 12px 12px, 12px 12px, 12px 12px;
              background-position: -9px 0px, -3px -6px, 3px 6px, -3px 0;
              left: 400px;
              top: 0;
              width: 375px;
              height: 812px;
            \\"
            data-uid=\\"scene-2-entity\\"
            data-label=\\"Same File App\\"
          >
            <div
              data-uid=\\"same-file-app-div\\"
              data-label=\\"Scene Thing\\"
              style=\\"
                position: relative;
                width: 100%;
                height: 100%;
                background-color: blue;
              \\"
              data-path=\\"storyboard-entity/scene-2-entity/same-file-app-entity:same-file-app-div\\"
            ></div>
          </div>
        </div>
      </div>
      "
    `)
  })
  it(`#1737 - Parser is broken for 'export const thing = "hello"'`, () => {
    const result = testCanvasRenderInlineMultifile(
      null,
      `import * as React from 'react'
import Utopia, {
  Scene,
  Storyboard,
  registerModule,
} from 'utopia-api'
import { App } from '/src/app'

export var storyboard = (
  <Storyboard data-uid='storyboard-entity'>
    <Scene
      data-label='Imported App'
      data-uid='scene-1-entity'
      style={{ position: 'absolute', left: 0, top: 0, width: 375, height: 812 }}
    >
      <App data-uid='app-entity' />
    </Scene>
  </Storyboard>
)`,
      {
        '/src/app.js': `import * as React from 'react'
import DefaultFunction, { Card, thing } from '/src/card.js'
export var App = (props) => {
  return (
    <div
      data-uid='app-outer-div'
      style={{
        position: 'relative',
        width: '100%',
        height: '100%',
        backgroundColor: '#FFFFFF',
      }}
    >
      <Card
        data-uid='card-instance'
        style={{
          position: 'absolute',
          left: 67,
          top: 0,
          width: 133,
          height: 300,
        }}
      />
      {thing}
      <DefaultFunction />
    </div>
  )
}`,
        '/src/card.js': `import * as React from 'react'
import { Rectangle } from 'utopia-api'
import { Spring } from 'non-existant-dummy-library'
export var Card = (props) => {
  return (
    <div
      data-uid='card-outer-div'
      style={{ ...props.style }}
    >
      <div
        data-uid='card-inner-div'
        style={{
          position: 'absolute',
          left: 0,
          top: 0,
          width: 50,
          height: 50,
          backgroundColor: 'red',
        }}
      />
      <Spring
        data-uid='card-inner-spring'
        style={{
          position: 'absolute',
          left: 100,
          top: 200,
          width: 50,
          height: 50,
          backgroundColor: 'blue',
        }}
      />
    </div>
  )
}

export const thing = 'hello'

export default function () {
  return <div>Default Function Time</div>
}`,
      },
    )
    expect(result).toMatchInlineSnapshot(`
      "<div style=\\"all: initial\\">
        <div
          id=\\"canvas-container\\"
          style=\\"position: absolute\\"
          data-utopia-valid-paths=\\"storyboard-entity storyboard-entity/scene-1-entity storyboard-entity/scene-1-entity/app-entity storyboard-entity/scene-1-entity/app-entity:app-outer-div storyboard-entity/scene-1-entity/app-entity:app-outer-div/card-instance storyboard-entity/scene-1-entity/app-entity:app-outer-div/d7f\\"
          data-utopia-root-element-path=\\"storyboard-entity\\"
        >
          <div
            data-utopia-scene-id=\\"storyboard-entity/scene-1-entity\\"
            data-path=\\"storyboard-entity/scene-1-entity\\"
            style=\\"
              overflow: hidden;
              position: absolute;
              background-color: rgba(255, 255, 255, 1);
              box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
              background-image: linear-gradient(
                  to bottom left,
                  #e7e7e7 25%,
                  transparent 25%
                ),
                linear-gradient(to bottom left, transparent 75%, #e7e7e7 75%),
                linear-gradient(to bottom right, #e7e7e7 25%, transparent 25%),
                linear-gradient(to bottom right, transparent 75%, #e7e7e7 75%);
              background-size: 12px 12px, 12px 12px, 12px 12px, 12px 12px;
              background-position: -9px 0px, -3px -6px, 3px 6px, -3px 0;
              left: 0;
              top: 0;
              width: 375px;
              height: 812px;
            \\"
            data-uid=\\"scene-1-entity\\"
            data-label=\\"Imported App\\"
          >
            <div
              data-uid=\\"app-outer-div\\"
              style=\\"
                position: relative;
                width: 100%;
                height: 100%;
                background-color: #ffffff;
              \\"
              data-path=\\"storyboard-entity/scene-1-entity/app-entity:app-outer-div\\"
            >
              <div
                data-uid=\\"card-outer-div\\"
                style=\\"
                  position: absolute;
                  left: 67px;
                  top: 0;
                  width: 133px;
                  height: 300px;
                \\"
                data-path=\\"storyboard-entity/scene-1-entity/app-entity:app-outer-div/card-instance:card-outer-div\\"
              >
                <div
                  data-uid=\\"card-inner-div\\"
                  style=\\"
                    position: absolute;
                    left: 0;
                    top: 0;
                    width: 50px;
                    height: 50px;
                    background-color: red;
                  \\"
                  data-path=\\"storyboard-entity/scene-1-entity/app-entity:app-outer-div/card-instance:card-outer-div/card-inner-div\\"
                ></div>
                <div
                  data-uid=\\"card-inner-spring\\"
                  style=\\"
                    position: absolute;
                    left: 100px;
                    top: 200px;
                    width: 50px;
                    height: 50px;
                    background-color: blue;
                  \\"
                  data-path=\\"storyboard-entity/scene-1-entity/app-entity:app-outer-div/card-instance:card-outer-div/card-inner-spring\\"
                ></div>
              </div>
              hello
              <div
                data-uid=\\"4cf\\"
                data-path=\\"storyboard-entity/scene-1-entity/app-entity:app-outer-div/d7f:4cf\\"
              >
                Default Function Time
              </div>
            </div>
          </div>
        </div>
      </div>
      "
    `)
  })
})
