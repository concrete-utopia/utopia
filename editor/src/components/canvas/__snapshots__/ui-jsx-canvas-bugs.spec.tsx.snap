// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiJsxCanvas #747 - DOM object constructor cannot be called as a function 1`] = `
"<div style=\\"all: initial\\">
  <div
    id=\\"canvas-container\\"
    style=\\"position: absolute\\"
    data-utopia-valid-paths=\\"sb sb/scene sb/scene/app\\"
    data-utopia-root-element-path=\\"sb\\"
  >
    <div
      data-utopia-scene-id=\\"sb/scene\\"
      data-path=\\"sb/scene\\"
      style=\\"
        overflow: hidden;
        position: absolute;
        background-color: rgba(255, 255, 255, 1);
        box-shadow: 0px 0px 1px 0px rgba(26, 26, 26, 0.3);
        left: 0;
        top: 0;
        width: 375px;
        height: 812px;
      \\"
      data-uid=\\"scene\\"
    >
      <div data-uid=\\"comment-root\\" data-path=\\"5c5~~~1:comment-root\\">hat</div>
    </div>
  </div>
</div>
"
`;

exports[`UiJsxCanvas #747 - DOM object constructor cannot be called as a function 2`] = `
Object {
  "sb/scene": Object {
    "attributeMetadatada": Object {},
    "componentInstance": false,
    "computedStyle": Object {},
    "element": Object {
      "type": "RIGHT",
      "value": Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": Object {
              "baseVariable": "App",
              "propertyPath": Object {
                "propertyElements": Array [],
              },
            },
            "props": Array [
              Object {
                "comments": Object {
                  "leadingComments": Array [],
                  "trailingComments": Array [],
                },
                "key": "data-uid",
                "type": "JSX_ATTRIBUTES_ENTRY",
                "value": Object {
                  "comments": Object {
                    "leadingComments": Array [],
                    "trailingComments": Array [],
                  },
                  "type": "ATTRIBUTE_VALUE",
                  "value": "app",
                },
              },
            ],
            "type": "JSX_ELEMENT",
            "uid": "app",
          },
        ],
        "name": Object {
          "baseVariable": "Scene",
          "propertyPath": Object {
            "propertyElements": Array [],
          },
        },
        "props": Array [
          Object {
            "comments": Object {
              "leadingComments": Array [],
              "trailingComments": Array [],
            },
            "key": "data-uid",
            "type": "JSX_ATTRIBUTES_ENTRY",
            "value": Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "type": "ATTRIBUTE_VALUE",
              "value": "scene",
            },
          },
          Object {
            "comments": Object {
              "leadingComments": Array [],
              "trailingComments": Array [],
            },
            "key": "style",
            "type": "JSX_ATTRIBUTES_ENTRY",
            "value": Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "type": "ATTRIBUTE_VALUE",
              "value": Object {
                "height": 812,
                "left": 0,
                "position": "absolute",
                "top": 0,
                "width": 375,
              },
            },
          },
        ],
        "type": "JSX_ELEMENT",
        "uid": "scene",
      },
    },
    "elementPath": Object {
      "parts": Array [
        Array [
          "sb",
          "scene",
        ],
      ],
      "type": "elementpath",
    },
    "globalFrame": null,
    "importInfo": Object {
      "type": "RIGHT",
      "value": Object {
        "originalName": "Scene",
        "path": "utopia-api",
        "variableName": "Scene",
      },
    },
    "isEmotionOrStyledComponent": false,
    "label": null,
    "localFrame": null,
    "specialSizeMeasurements": Object {
      "clientHeight": 0,
      "clientWidth": 0,
      "coordinateSystemBounds": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "display": "initial",
      "flexDirection": null,
      "globalContentBox": null,
      "htmlElementName": "div",
      "immediateParentBounds": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "immediateParentProvidesLayout": true,
      "layoutSystemForChildren": "flow",
      "margin": Object {},
      "naturalHeight": null,
      "naturalWidth": null,
      "offset": Object {
        "x": 0,
        "y": 0,
      },
      "padding": Object {},
      "parentFlexDirection": null,
      "parentLayoutSystem": "flow",
      "position": "static",
      "providesBoundsForChildren": false,
      "renderedChildrenCount": 0,
      "usesParentBounds": false,
    },
  },
  "sb/scene/app": Object {
    "attributeMetadatada": Object {},
    "componentInstance": false,
    "computedStyle": Object {},
    "element": Object {
      "type": "RIGHT",
      "value": Object {
        "children": Array [],
        "name": Object {
          "baseVariable": "App",
          "propertyPath": Object {
            "propertyElements": Array [],
          },
        },
        "props": Array [
          Object {
            "comments": Object {
              "leadingComments": Array [],
              "trailingComments": Array [],
            },
            "key": "data-uid",
            "type": "JSX_ATTRIBUTES_ENTRY",
            "value": Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "type": "ATTRIBUTE_VALUE",
              "value": "app",
            },
          },
        ],
        "type": "JSX_ELEMENT",
        "uid": "app",
      },
    },
    "elementPath": Object {
      "parts": Array [
        Array [
          "sb",
          "scene",
          "app",
        ],
      ],
      "type": "elementpath",
    },
    "globalFrame": null,
    "importInfo": Object {
      "type": "LEFT",
      "value": "NOT_IMPORTED",
    },
    "isEmotionOrStyledComponent": false,
    "label": null,
    "localFrame": null,
    "specialSizeMeasurements": Object {
      "clientHeight": 0,
      "clientWidth": 0,
      "coordinateSystemBounds": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "display": "initial",
      "flexDirection": null,
      "globalContentBox": null,
      "htmlElementName": "div",
      "immediateParentBounds": Object {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "immediateParentProvidesLayout": true,
      "layoutSystemForChildren": "flow",
      "margin": Object {},
      "naturalHeight": null,
      "naturalWidth": null,
      "offset": Object {
        "x": 0,
        "y": 0,
      },
      "padding": Object {},
      "parentFlexDirection": null,
      "parentLayoutSystem": "flow",
      "position": "static",
      "providesBoundsForChildren": false,
      "renderedChildrenCount": 0,
      "usesParentBounds": false,
    },
  },
}
`;
